flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,418 @@
+Cadence Mutability Restrictions
+
+| Status        | (Proposed)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [703](https://github.com/onflow/flow/pull/703)       |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2021-12-08                                           |
+
+## Objective
+
+This proposed change would limit the scopes in which the fields of composite types
+like contracts, structs, and resources can be mutated. Instead of allowing array 
+and dictionary fields to be modified in any scope where the field can be read,
+Cadence would issue a type error. These fields would instead be only modifiable 
+in the current declaration scope, as well as inner scopes of that scope. 
+
+## Motivation
+
+Accidentally exposing a mutable variable to external consumers of a contract is currently a 
+large potential footgun standing in the way of a release of a stable, trustless version of
+Cadence. Developers may declare a "constant" field on their contract with `pub let`, intending
+that the field only be readable to transactions and other contracts, and unintentially allow
+other code to add or remove elements from a dictionary or array stored in that field. Consider this code:
+
+```
+pub contract Foo {
+	pub let x : [Int]
+
+	init() {
+	    self.x = []
+	}
+}
+
+// in some external code importing Foo
+pub fun bar() {
+	Foo.x.append(1)
+}
+```
+
+Currently Cadence does not warn against this, or prevent a developer from writing this code, even
+though, depending on what is stored in `x`, this could be unsafe. 
+
+## User Benefit
+
+This change makes it less likely for a user to expose a value from a contract or other
+publicly accessible structure that is not intended to be mutable, but can be modified 
+by consumers of their contract regardless. This is important for a future stable, 
+permissionless version of Cadence, where users do not need to have their smart
+contracts reviewed in order to deploy them to Flow. By removing one possible way that 
+users could write unsafe smart contracts, we make it easier for everyone to trust the
+correctness of contracts deployed to Flow. 
+
+## Design Proposal
+
+### Explanation 
+
+Cadence controls where and to what extent variables can be read and written using a combination of
+access modifiers and declaration kinds, as described [here](https://docs.onflow.org/cadence/language/access-control/).
+Of note is that the `let` kind does not allow fields to be written to in any scope, whereas `var` allows them
+to be written in the "Current and Inner" scopes; that is, the scope in which the field was declared, and any scopes
```

Comment:
"written to" sounds like mutation is not allowed, but in fact it is, and this is exactly what this proposal is trying to introduce.

Maybe reword to something like "reassigned", etc.